File: sliding-window/NumberOfSubString.java
public class NumberOfSubString {
    public static int numberOfSubstrings(String s) {
        int n = s.length();
        int left = 0;
        int[] cnt = new int[3];
        int ans = 0;

        for (int right = 0; right < n; right++) {
            char c = s.charAt(right);
            if (c >= 'a' && c <= 'c') cnt[c - 'a']++;

            while (cnt[0] > 0 && cnt[1] > 0 && cnt[2] > 0) {
                ans += n - right;
                char lc = s.charAt(left++);
                if (lc >= 'a' && lc <= 'c') cnt[lc - 'a']--;
            }
        }

        return ans;
    }

    public static void main(String[] args) {
        System.out.println(numberOfSubstrings("abcabc"));
    }
}

File: sliding-window/number_of_substring.js
function numberOfSubstrings(str) {
  let n = str.length;
  let left_ptr = 0;
  let right_ptr = 0;
  let cnt = 0;
  const mpp = new Map();

  while (right_ptr < n) {
    mpp.set(str[right_ptr], (mpp.get(str[right_ptr]) || 0) + 1);

    while (mpp.size === 3) {
      cnt += n - right_ptr;

      mpp.set(str[left_ptr], mpp.get(str[left_ptr]) - 1);
      if (mpp.get(str[left_ptr]) === 0) {
        mpp.delete(str[left_ptr]);
      }
      left_ptr++;
    }
    right_ptr++;
  }
  return cnt;
}

const cnt = numberOfSubstrings("abcabc");
console.log(cnt);
